### ISSUE TITLE
WireMock with JUnit 5

### ISSUE THREAD

**Issue by Wonha** [tomakehurst/wiremock#684]
In JUnit 5, `@Rule` and `@ClassRule` is superseded by `@ExtendWith`, which requires extension class.
And it seems like extension for WireMock is not yet supported.
Is there any alternative way to use WireMock with stubbing and verification via JUnit 5?

**Comment by tomakehurst** [comment]
You can programmatically start and stop `WireMockServer`. That's all the rule is doing really.

I'll mark this as a feature request as JUnit 5 support should start to be a priority soon.

**Comment by Wonha** [comment]
Thank you for your comment, and marking.
Actually, I tried to use WireMock with JUnit, and there was no other way to do it without using JUnit 4.x Rule.
Thanks !

**Comment by tomakehurst** [comment]
You can use  `@Before` and `@After` annotated methods to start and stop a `WireMockServer` in JU4.

**Comment by tomakehurst** [comment]
Some of WireMock's own acceptance tests actually do this in their base class, so might be worth checking that out if you're not sure what I mean.

**Comment by Wonha** [comment]
Thanks for your kind comment. I'll check it.

**Comment by lanwen** [comment]
Please look at https://github.com/lanwen/wiremock-junit5 (will be available in maven central soon)

**Comment by sta-szek** [comment]
@lanwen nice! Could you do this as class level fields instead of (or additionally) method parameters?

**Comment by lanwen** [comment]
@sta-szek Like a class rule?

**Comment by sta-szek** [comment]
@lanwen exactly - just to simplify as much as possible.

**Comment by Wonha** [comment]
@lanwen thanks for the nice code!

**Comment by JensPiegsa** [comment]
https://github.com/JensPiegsa/wiremock-extension

**Comment by lanwen** [comment]
What the difference?

**Comment by JensPiegsa** [comment]
It's an approach that uses annotated fields. See [ExampleTest](https://github.com/JensPiegsa/wiremock-extension/blob/master/src/test/java/com/github/jenspiegsa/mockitoextension/ExampleTest.java)

**Comment by goyalashish** [comment]
@lanwen can you help me how to wiremock-junit5 for custom port.
As in your code, as per my understanding, we can get any random port, but if you want a specific port for wiremock, can you specify how to do that.

void function(@WiremockResolver.Wiremock WireMockServer server) throws IOException {
        ....
}

but in this, i can't set server port for which server need to start, 
I was checking the code of wiremockResolver, this was the code,
this.server = new WireMockServer(((WiremockConfigFactory)mockedServer.factory().newInstance()).create());
i guess, due to this.. i am unable to specify some specific port.

Is there is any way to do that ?
Please unicast me

**Comment by ghilainm** [comment]
> You can programmatically start and stop `WireMockServer`. That's all the rule is doing really.
> 
> I'll mark this as a feature request as JUnit 5 support should start to be a priority soon.

Any news about this?

**Comment by jmayday** [comment]
@ghilainm for me rewriting rule to Junit5 extension looks like this: https://github.com/jmayday/wiremock/commit/908cbb20d19cf7dcabcca3de28a2712fd895e472 (I'm not sure if it satisfies requirements for pull request though, @tomakehurst?)

As server starts on random port, we need to build also a webclient with correct url. It might be achieved in beforeAll method:
```
  @BeforeAll
  static void setup(@ServerUrl String serverUrl) {
    // build webclient using server url
  }
```

**Comment by tomakehurst** [comment]
I'm a bit reticent about making WireMock dependent on both JUnit 4 and 5. Dropping 4 isn't an option at the moment as it's still pretty widely used.

Do you think there's much disadvantage in adding JUnit 5 support via a separate library?

**Comment by behrangsa** [comment]
> Do you think there's much disadvantage in adding JUnit 5 support via a separate library?

Yes, an official but optional library should be quite helpful.

**Comment by luolong** [comment]
What about splitting Wiremock library in two -- `wiremock-core` and `wiremock-junit4` and adding `wiremock-jupiter` for JUnit 5 support?

The `wiremock` library could then simply have a dependency on core and junit4 for the sake of legacy users?

**Comment by tomakehurst** [comment]
@luolong that doesn't sound much different from what we have now, if you consider @lanwen's library to be the jupiter version.

**Comment by t1** [comment]
Having an 'official' Wiremock JUnit5 extension would be valuable. I actually expected to find a JUnit 5 extension when I have a JUnit 4 rule. And I think an optional dependency on JUnit 5 would be no problem at all, would it?

**Comment by luolong** [comment]
@tomakehurst Sure, I realize that this is not so much different from what you have now. This is by design.

The only difference is that JUnit 4 dependency is explicitly divorced from the core dependency.
The separate modules could even share the same repository and release schedule.

Should not be difficult to achieve.

**Comment by tomakehurst** [comment]
I'm coming around to the idea of making both JUnit 4 and 5 optional/provided dependencies. I'm going to look at this (with the usual caveats about being busy and it not happening quickly).

**Comment by blocha** [comment]
Any update on this?

**Comment by JcMinarro** [comment]
Any update on this?
Currently, I am using @JensPiegsa implementation, but would be great to have an "official" artifact from Wiremock repository.

**Comment by fransflippo** [comment]
@tomakehurst Is this something any of us could help with? I can imagine splitting out the junit4 dependencies can be a hairy task with some architectural impact, so I would understand if you want to do it yourself since Wiremock is your baby :-) 
But otherwise, let us know!

**Comment by blocha** [comment]
@fransflippo @tomakehurst I would recommend using:

@AutoConfigureWireMock from our friends from spring's 'org.springframework.cloud:spring-cloud-contract-wiremock' - it's working great with Junit5 and you can customize it it you want :)

**Comment by fransflippo** [comment]
> @fransflippo @tomakehurst I would recommend using:
> 
> @AutoConfigureWireMock from our friends from spring's 'org.springframework.cloud:spring-cloud-contract-wiremock' - it's working great with Junit5 and you can customize it it you want :)

Thanks @blocha , will take a look!

**Comment by tomakehurst** [comment]
@fransflippo the JUnit 4 dependency is already optional, so it might be quite straightforward to do something similar with JUnit 5.

I'd definitely welcome a PR to add this if you'd like to contribute.

I like the approach @lanwen has taken i.e. creating [JUnit 5 extensions](https://github.com/lanwen/wiremock-junit5) to inject WireMock classes as test method parameters, although I'd be inclined to inject an instance of `WireMock` rather than `WireMockServer`.

**Comment by jglick** [comment]
Canned rules are sometimes more trouble than they are worth. My workaround:

```java
private WireMockServer wireMockServer;

@BeforeEach public void proxyToWireMock() {
    wireMockServer = new WireMockServer(WireMockConfiguration.options()/*.port(etc)*/);
    wireMockServer.start();
}

@AfterEach public void noMoreWireMock() {
    wireMockServer.stop();
    wireMockServer = null;
}
```

**Comment by fdw** [comment]
There is one aspect the rules were quite good at that are not possible anymore with @lanwen's or @JensPiegsa's implementations: Having more than one mock server. We have one server that talks to several others, and having different mock servers made the tests more readable. Theoretically, we can only use one mock server, but then it is hard to understand where each request is supposed to be sent.

We could also do this like @jglick suggests, but this scales badly, too.

**Comment by JensPiegsa** [comment]
@fdw Of course, there is support for multiple server instances. Please have a look at the [example](https://github.com/JensPiegsa/wiremock-extension/blob/master/src/test/java/com/github/jenspiegsa/wiremockextension/ExampleTest.java#L124).

**Comment by tomakehurst** [comment]
@fdw this is a very good point, and I think any solution should attempt to cleanly support multiple instances. I haven't looked at JUnit5 extension architecture at all closely, so I'm open to suggestion as to how this could be done.

**Comment by fdw** [comment]
@JensPiegsa You are right, I hadn't seen that example. Thank you :)

**Comment by jterhune** [comment]
If it helps anyone, we are doing multiple wiremock servers with this single extension class:

```
class WireMockExtension extends WireMockServer implements BeforeEachCallback, AfterEachCallback {

    @Override
    public void beforeEach(ExtensionContext context) {
        this.start();
    }

    @Override
    public void afterEach(ExtensionContext context) {
        this.stop();
        this.resetAll();
    }
}
```

Use it with multiple servers in a test like this:
```
        @RegisterExtension
        WireMockExtension mockServer1 = new WireMockExtension(8081)

        @RegisterExtension
        WireMockExtension mockServer2 = new WireMockExtension(8082)
```

**Comment by darmbrust** [comment]
I'm getting this odd error, if I try to use wiremock with JUnit 5... 

	@AfterAll
	public void after() throws Exception {
		super.tearDown();
		wireMockRule.shutdown();
	}

error: cannot access MethodRule
[ERROR]   class file for org.junit.rules.MethodRule not found

It only happens if I try to do the shutdown.  Is there another way I can shut down the server without triggering this error?

**Comment by darmbrust** [comment]
ahh, how silly.  I had copied and pasted some snippits from demos, and was trying to do new WireMockRule(8089) instead of new WireMockServer(8089), which obviously hits a class that has dependencies on the Junit 4 API.

**Comment by sergey-morenets** [comment]
@jterhune In my case I also had to add WireMock.configureFor to your solution:

```
 @Override
    public void beforeEach(ExtensionContext context) {
        this.start();
        WireMock.configureFor("localhost", port());
    }
```

Otherwise WireMock used port 8080 for admin mappings and can't connect to it the tests.

**Comment by Patouche** [comment]
Hello, 

If you're ok with that, I can try to add this feature in WireMock.

Recently, I have setup WireMock in a project at work. I was a little bit surprised that this is not already part of WireMock and we need to add an other library to add this kind of extension.

First at all, I took a look of some repositories such as :

* https://github.com/lanwen/wiremock-junit5 
* https://github.com/JensPiegsa/wiremock-extension 
* https://github.com/phxql/wiremock-junit5
* ...

It gave my some inspiration but I found that each extensions had some missing features. For example, none of these 3 JUnit extensions allow you to have an alternative between constructor injection, field injection or method injection.

For the configuration, somes extension didn't seem very clear to me. However, there are great ideas in each of these community extensions.

I can try to put all ideas together and maybe to add new one to have an official JUnit 5 extension for WireMock.

What do you think of that ?

**Comment by tomakehurst** [comment]
If you're willing to work on this, that would be awesome. JUnit 5 is very high on my priority list, but I've just not had time to devote to it.

Happy to provide support and design input as you need it.

**Comment by rieckpil** [comment]
Great to see @Patouche step in to add a JUnit Jupiter extension 🚀 

In the meantime and for those who are looking for a solution to start WireMock before all test methods (as far as I can see it, no community exertion allows this as they are implementing the `BeforeEachCallback`) especially when writing test that involve a SpringTestContext where one wants to set e.g. the base URL prior to starting the application, I used the following setup over the past months:

```
public class WireMockInitializer 
  implements ApplicationContextInitializer<ConfigurableApplicationContext> {
 
  @Override
  public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
    WireMockServer wireMockServer = new WireMockServer(new WireMockConfiguration().dynamicPort());
    wireMockServer.start();
 
    configurableApplicationContext
      .getBeanFactory()
      .registerSingleton("wireMockServer", wireMockServer);
 
    configurableApplicationContext.addApplicationListener(applicationEvent -> {
      if (applicationEvent instanceof ContextClosedEvent) {
        wireMockServer.stop();
      }
    });
 
    TestPropertyValues
      // override any base URL of your HTTP clients here
      .of(Map.of("todo_base_url", "http://localhost:" + wireMockServer.port()))
      .applyTo(configurableApplicationContext);
  }
}
```

With the initializer above you'll start WireMock as part of the bootstrapping phase of the Spring Context, register it as a Spring Bean and can override base URLs for your HTTP clients.

Next, you need to activate/include this initializer with `@ContextConfiguration`:

```
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ContextConfiguration(initializers = {WireMockInitializer.class})
class TodoControllerIT {
 
  @Autowired
  private WireMockServer wireMockServer;
 
  @AfterEach
  public void afterEach() {
    this.wireMockServer.resetAll();
  }
}
```

You can find a detailed explanation as part of this [blog post](https://rieckpil.de/spring-boot-integration-tests-with-wiremock-and-junit-5/).

**Comment by p7u** [comment]
> Hello,
> 
> If you're ok with that, I can try to add this feature in WireMock.
> 
> Recently, I have setup WireMock in a project at work. I was a little bit surprised that this is not already part of WireMock and we need to add an other library to add this kind of extension.
> 
> First at all, I took a look of some repositories such as :
> 
> * https://github.com/lanwen/wiremock-junit5
> * https://github.com/JensPiegsa/wiremock-extension
> * https://github.com/phxql/wiremock-junit5
> * ...
> 
> It gave my some inspiration but I found that each extensions had some missing features. For example, none of these 3 JUnit extensions allow you to have an alternative between constructor injection, field injection or method injection.
> 
> For the configuration, somes extension didn't seem very clear to me. However, there are great ideas in each of these community extensions.
> 
> I can try to put all ideas together and maybe to add new one to have an official JUnit 5 extension for WireMock.
> 
> What do you think of that ?

So, any news?

**Comment by beatngu13** [comment]
@tomakehurst I just gave it a try. Unfortunately, the beta version is not available on central and [JitPack has issues](https://jitpack.io/com/github/tomakehurst/wiremock/2.31.0-beta/build.log) building the project. So I downloaded the [JAR](https://github.com/tomakehurst/wiremock/releases/download/2.31.0-beta/wiremock-jre8-standalone-2.31.0-beta.jar) and added it to my local repo:

```text
mvn install:install-file \
        -Dfile=wiremock-jre8-standalone-2.31.0-beta.jar \     
        -DgroupId=com.github.tomakehurst \
        -DartifactId=wiremock-jre8 \
        -Dversion=2.31.0-beta \
        -Dpackaging=jar
```

Works like a charm, almost like a drop-in replacement – great work!

Side note: it's JUnit 5, so you can remove the `public` modifiers from the [docs](https://github.com/tomakehurst/wiremock/blob/junit-jupiter/docs-v2/_docs/junit-jupiter.md).

**Comment by tomakehurst** [comment]
Thanks for the feedback @beatngu13!

It's on my list to get JitPack working so should be easier with future betas.

**Comment by beatngu13** [comment]
@tomakehurst one suggestion: maybe use `Options#DEFAULT_PORT` (i.e. 8080) [here](https://github.com/tomakehurst/wiremock/blob/0da7a6f3acd5dcc33f894da41c1d8f4f1639ce0b/src/main/java/com/github/tomakehurst/wiremock/junit5/WireMockExtension.java#L23) in the extension, rather than defaulting to a random port. Then, the extension IMHO better aligns with WireMock's default behavior and the existing JUnit 4 rules; also users don't have to "fiddle" with `@RegisterExtension` and can use the more popular `@ExtendWith` annotation. (Convention over configuration, so to say…)

**Comment by tomakehurst** [comment]
I agree it's less surprising to default to 8080, but OTOH it prevents you from running your tests in parallel without switching everything over to the programmatic form, which you tend to want to do only when you've got 100s of tests and your build is starting to feel slow.

Also, defaulting to the random port nudges developers to use the WireMock's base URL methods rather than `"http://localhost:8080" + myPath`, which seems like better practice to me.

I'm genuinely in two minds though, not least because I can see myself spending the next 5 years telling every other mailing list poster about the random port.

**Comment by tomakehurst** [comment]
There's another possibility, arguably a middle way.

I've been thinking it'd be good to support automatic JVM proxy setup in a similar fashion to Hoverfly, whereby forward proxying is enabled and the JVM system properties for proxy config are set to point to it so that you can mock multiple domains in a single instance.

I wouldn't want to turn this on by default, but it'd be nice to have it available declaratively so I was wondering about creating a subclass to the extension that turns this option on e.g.

```java
@ExtendWith(WireMockExtensionWithJvmProxy.class)
```

Perhaps we could also have something like:

```java
@ExtendWith(WireMockExtensionRandomPort.class)
```

or if random port is kept as the the default:

```java
@ExtendWith(WireMockExtensionPort8080.class)
```

Thoughts?

**Comment by beatngu13** [comment]
Ah, I see. Using a random port definitely makes sense when executing the tests concurrently. You also know your user base way better than I do and what default makes more sense.

Regarding `WireMockExtensionRandomPort` and `WireMockExtensionPort8080`, you could also do something like:

```java
@Target( ElementType.TYPE )
@Retention(RetentionPolicy.RUNTIME)
@ExtendWith(WireMockExtension.class)
public @interface WireMockTest {

    // null means random port
    Integer httpPort;

    // more core config properties

}
```

Which can be used like:

```java
@WireMockTest
class DeclarativeWireMockTest { }
```

If a static port is desired:


```java
@WireMockTest(httpPort = 8080)
class DeclarativeWireMockTest { }
```

**Comment by tomakehurst** [comment]
Yes, that's much nicer isn't it. I'll go with that.

**Comment by tomakehurst** [comment]
I've updated the branch to include this ☝️ now.
https://github.com/tomakehurst/wiremock/blob/junit-jupiter/docs-v2/_docs/junit-jupiter.md

I'm still trying to figure out a better way to publish a beta build than putting the standalone JAR on Releases. So far:
1) JitPack won't work because there are build steps that rely on NPM.
2) GitHub Packages won't work because it won't let you make Maven repos public (unauthenticated).
3) Sonatype's OSS snapshots repo won't work because it's giving me unexplained 400 errors when I try to publish to it.

If anyone has any other ideas I'm all ears...

**Comment by szpak** [comment]
> 3. Sonatype's OSS snapshots repo won't work because it's giving me unexplained 400 errors when I try to publish to it.

Seems to be the client side problem. I have snapshots publishing in a few projects and it used to work fine.

Do you have some CI logs to take a look (and configuration in a branch)?

**Comment by tomakehurst** [comment]
@szpak there's nothing useful in the logs really, just `Could not PUT 'https://oss.sonatype.org/content/repositories/snapshots/com/github/tomakehurst/wiremock-jre8/2.31.0-beta-2/wiremock-jre8-2.31.0-beta-2.jar'. Received status code 400 from server: Bad Request`.

Do you have a Gradle build file with working sonatype snapshot publishing you could share so I can compare?

**Comment by tomakehurst** [comment]
OK, looks like the issue is that you have to actually have `SNAPSHOT` in the version (duh!).

I've pushed a `2.31.0-SNAPSHOT` to the Sonatype snapshots repo now for those wanting to try it out. You'll need to use https://oss.sonatype.org/content/repositories/snapshots/ as the repo URL in your build.

**Comment by szpak** [comment]
> OK, looks like the issue is that you have to actually have SNAPSHOT in the version (duh!).

Yup, they have separate repos for [snapshot](https://oss.sonatype.org/content/repositories/snapshots/com/github/tomakehurst/wiremock-jre8/) and release versions and are strict about content. I'm glad you solved it!

**Comment by bmorris591** [comment]
Thanks for this! Just playing with the new SNAPSHOT build - looks like the noargs constructor on `WireMockExtension` is `private` and the [example docs][1] have `newInstance()` calls to get an instance programatically.

It might be good to make the API more "friendly" for programmatic use? Perhaps include some `static` factory methods or a builder? 

[1]: https://github.com/tomakehurst/wiremock/blob/junit-jupiter/docs-v2/_docs/junit-jupiter.md

**Comment by bmorris591** [comment]
There seems to be some issue with the setup of Gradle for the JUnit BOM. If you look at the [pom][1] you can see the following 

```xml
<dependency>
    <groupId>org.junit</groupId>
    <artifactId>junit-bom</artifactId>
    <version>5.7.1</version>
    <scope>runtime</scope>
</dependency>
```
i.e. it adds the BOM as the runtime dependency rather than in the `dependencyManagement` section and the scope is `runtime` not `import`. 

Not sure if this is a Gradle bug, but it makes Gradle builds unhappy (running Gradle 7.2)

```
> Could not resolve all files for configuration ':testRuntimeClasspath'.
   > Could not resolve org.junit:junit-bom:5.7.1.
     Required by:
         project : > com.github.tomakehurst:wiremock-jre8:2.31.0-SNAPSHOT:20210829.210337-1
      > No matching variant of org.junit:junit-bom:5.7.2 was found. The consumer was configured to find a runtime of a library compatible with Java 16, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
          - Variant 'apiElements' capability org.junit:junit-bom:5.7.2:
              - Incompatible because this component declares an API of a platform and the consumer needed a runtime of a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 16)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
          - Variant 'enforcedApiElements' capability org.junit:junit-bom-derived-enforced-platform:5.7.2:
              - Incompatible because this component declares an API of an enforced platform and the consumer needed a runtime of a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 16)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
          - Variant 'enforcedRuntimeElements' capability org.junit:junit-bom-derived-enforced-platform:5.7.2 declares a runtime of a component:
              - Incompatible because this component declares an enforced platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 16)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
          - Variant 'runtimeElements' capability org.junit:junit-bom:5.7.2 declares a runtime of a component:
              - Incompatible because this component declares a platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 16)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')

```

This is because Gradle "knows" that the BOM is a "platform" but it's being added as a dependency.

Excluding the dependency from Wiremock makes things happy

```groovy
    testImplementation("com.github.tomakehurst:wiremock-jre8:2.31.0-SNAPSHOT") {
        exclude(group = "org.junit", module = "junit-bom")
    }
```

[1]: https://oss.sonatype.org/service/local/repositories/snapshots/content/com/github/tomakehurst/wiremock-jre8/2.31.0-SNAPSHOT/wiremock-jre8-2.31.0-20210829.210337-1.pom

**Comment by tomakehurst** [comment]
Thanks for reporting this.

Would you mind sharing your Gradle build file (or enough of it) so I can reproduce this?

Also, does this happen with Java versions < 16?

**Comment by bmorris591** [comment]
It's publicly available on the competing code hosting platform - https://gitlab.com/bmorris591/volvooncall-mysql

I've pushed up a [branch without the exclusion][1] to show the error, mainline [has the exclusion][2] and compiles happily.

[CI/CD shows the error][3]

[1]: https://gitlab.com/bmorris591/volvooncall-mysql/-/tree/spike/wiremock-junit5
[2]: https://gitlab.com/bmorris591/volvooncall-mysql/-/blob/master/build.gradle.kts#L137
[3]: https://gitlab.com/bmorris591/volvooncall-mysql/-/jobs/1549999016

**Comment by tomakehurst** [comment]
Thanks for the detailed info @bmorris591!

I think I've figured out the solution, so I'll push another snapshot tomorrow.

**Comment by elect86** [comment]
@bmorris591 I could fix the Gradle synchronization with `::exclude`, but if I try to run a test, I still get:
> e: /home/elect/IdeaProjects/gradle-download-task/src/test/kotlin/de/undercouch/gradle/tasks/download/testBases.kt: (89, 22): Cannot access 'org.junit.rules.MethodRule' which is a supertype of 'com.github.tomakehurst.wiremock.junit.WireMockClassRule'. Check your module classpath for missing or conflicting dependencies
e: /home/elect/IdeaProjects/gradle-download-task/src/test/kotlin/de/undercouch/gradle/tasks/download/testBases.kt: (89, 22): Cannot access 'org.junit.rules.TestRule' which is a supertype of 'com.github.tomakehurst.wiremock.junit.WireMockClassRule'. Check your module classpath for missing or conflicting dependencies
e: /home/elect/IdeaProjects/gradle-download-task/src/test/kotlin/de/undercouch/gradle/tasks/download/testBases.kt: (94, 22): Cannot access 'org.junit.rules.MethodRule' which is a supertype of 'com.github.tomakehurst.wiremock.junit.WireMockClassRule'. Check your module classpath for missing or conflicting dependencies
e: /home/elect/IdeaProjects/gradle-download-task/src/test/kotlin/de/undercouch/gradle/tasks/download/testBases.kt: (94, 22): Cannot access 'org.junit.rules.TestRule' which is a supertype of 'com.github.tomakehurst.wiremock.junit.WireMockClassRule'. Check your module classpath for missing or conflicting dependencies

**Comment by tomakehurst** [comment]
New 2.31.0-SNAPSHOT pushed. This seems to fix the issue in your project @bmorris591.

**Comment by bmorris591** [comment]
That got a green build when I reran the spike linked above. Mainline build has also completed successfully after merge.

Thanks!

**Comment by alexsapps** [comment]
Seems fixed and this works for me!
http://wiremock.org/docs/junit-jupiter/

At first I thought we still had to make our own extensions because this issue was still open. Posting the link above so others can find it too.

**Comment by tomakehurst** [comment]
Good point, closed!
